### PR TITLE
Added payload convention tests and fix the conversion bug

### PR DESCRIPTION
**Pull request by tmahmood-microsoft**

### CODE REVIEW

File: iothub/device/src/Utilities/ObjectConversionHelper.cs [hunk 1]
```diff
@@ -20,7 +20,7 @@ internal static bool TryCastOrConvert<T>(object objectToCastOrConvert, PayloadCo
             {
                 // We'll serialize the object back to JSON using the user's configured payload convention
                 // and then to the type of their choosing with the same payload convention.
-                value = payloadConvention.GetObject<T>(payloadConvention.GetObjectBytes(objectToCastOrConvert));
+                value = payloadConvention.GetObject<T>(DefaultPayloadConvention.Instance.GetObjectBytes(objectToCastOrConvert));
```
Comment (abhipsaMisra) [inline iothub/device/src/Utilities/ObjectConversionHelper.cs:23]: The comment states that both of these operations are expected to use the user's payload convention.
Your change is updating one of the conversions to always use Newtonsoft.

Can you elaborate on this change?

Comment (drwill-ms) [reply iothub/device/src/Utilities/ObjectConversionHelper.cs:23]: The comment needs updating.

The twin was deserialized with the default payload converter. So using the user's (when it is STJ) will fail to create object bytes that are compatible. Presumably the object is a JObject or something, right?

So we need to use the same payload convention used to deserialize the twin to serialize it back, and then use the user's one to deserialize.

File: iothub/device/src/Utilities/ObjectConversionHelper.cs
Comment (drwill-ms) [inline iothub/device/src/Utilities/ObjectConversionHelper.cs:1]: It occurs to me this if this is used in any code path where the SDK didn't deserialize using the default payload convention, we may have created another bug by hard coding the Newtonsoft here.

Can you validate other operations as well with E2E tests? Direct method, Incoming message.

Comment (abhipsaMisra) [reply iothub/device/src/Utilities/ObjectConversionHelper.cs:1]: That's my thought as well. If the intention is to do something that is specific to how twins are treated in our SDK then I'd recommend putting that code in the twin class and not in a "generic" helper. That way someone is less likely to reference it unintentionally in a use-case that it isn't intended for.

Comment (abhipsaMisra) [reply iothub/device/src/Utilities/ObjectConversionHelper.cs:1]: If the intention is to broadly use the default convention for the first step of the conversion then I'd recommend renaming this method appropriately - TryCastOrConvertFromJObject<T>

Comment (drwill-ms) [reply iothub/device/src/Utilities/ObjectConversionHelper.cs:1]: There's no intention. There's the reality we're now perceiving and realizing that the existing code wasn't working for STJ.

Comment (tmahmood-microsoft) [reply iothub/device/src/Utilities/ObjectConversionHelper.cs:1]: Added more e2e tests for direct method and incoming message

File: iothub/device/src/Utilities/ObjectConversionHelper.cs (hunk 1 above)
Comment (tmahmood-microsoft) [reply iothub/device/src/Utilities/ObjectConversionHelper.cs:23]: Updated the comment

File: iothub/device/src/DirectMethod/DirectMethodRequest.cs [hunk 2]
```diff
@@ -121,7 +121,7 @@ public bool TryGetPayload<T>(out T payload)
 
             try
             {
-                payload = PayloadConvention.GetObject<T>(Payload);
+                payload = DefaultPayloadConvention.Instance.GetObject<T>(Payload);
```
Comment (drwill-ms) [inline iothub/device/src/DirectMethod/DirectMethodRequest.cs:124]: This should be done with the user's selected payload convention. Presumably you have an e2e test with STJ selected using a custom class with only STJ attributes, right? It should fail this scenario.

File: iothub/device/src/Messaging/IncomingMessage.cs [hunk 3]
```diff
@@ -168,7 +168,7 @@ public bool TryGetPayload<T>(out T payload)
 
             try
             {
-                payload = PayloadConvention.GetObject<T>(_payload);
+                payload = DefaultPayloadConvention.Instance.GetObject<T>(_payload);
```
Comment (drwill-ms) [inline iothub/device/src/Messaging/IncomingMessage.cs:171]: Same here

File: e2e/Tests/iothub/device/DeviceCustomProperty.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace Microsoft.Azure.Devices.E2ETests.Twins
+{
+    internal class DeviceCustomProperty
+    {
+        [JsonPropertyName("customProperty")]
+        public string CustomProperty { get; set; }
+
+        [JsonPropertyName("guid")]
+        public string Guid { get; set; }
```
Comment (drwill-ms) [inline e2e/Tests/iothub/device/DeviceCustomProperty.cs:14]:
```suggestion
        [JsonPropertyName("string")]
        public string StringProperty { get; set; }

        [JsonPropertyName("guid")]
        public string GuidProperty { get; set; }
```

File: e2e/Tests/iothub/device/PayloadConventionE2eTests.cs [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+            using var cts = new CancellationTokenSource(s_testTimeout);
+            CancellationToken ct = cts.Token;
+            PayloadConvention convention = SystemTextJsonPayloadConvention.Instance;
+            var options = new IotHubClientOptions(new IotHubClientAmqpSettings())
+            {
+                PayloadConvention = convention
+            };
+            await using TestDevice testDevice = await TestDevice.GetTestDeviceAsync(s_devicePrefix).ConfigureAwait(false);
+            await using var deviceClient = new IotHubDeviceClient(testDevice.ConnectionString, options);
+            await deviceClient.OpenAsync().ConfigureAwait(false);
```
Comment (drwill-ms) [inline e2e/Tests/iothub/device/PayloadConventionE2eTests.cs:37]: consider using `testDevice.OpenWithRetryAsync`

File: e2e/Tests/iothub/device/PayloadConventionE2eTests.cs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+                    msg.TryGetPayload(out StjCustomPayload actual).Should().BeTrue();
+                    actual.StringProperty.Should().Be(requestPayload.StringProperty);
+                    actual.GuidProperty.Should().Be(requestPayload.GuidProperty);
+                    messageReceived.TrySetResult(true);
+                    var response = new DirectMethodResponse(200)
+                    {
+                        Payload = responsePayload,
+                    };
+                    return Task.FromResult(response);
+                },
```
Comment (drwill-ms) [inline e2e/Tests/iothub/device/PayloadConventionE2eTests.cs:144]:
```suggestion
                    {
                        msg.TryGetPayload(out StjCustomPayload actual).Should().BeTrue();
                        actual.StringProperty.Should().Be(requestPayload.StringProperty);
                        actual.GuidProperty.Should().Be(requestPayload.GuidProperty);
                        messageReceived.TrySetResult(true);
                        var response = new DirectMethodResponse(200)
                        {
                            Payload = responsePayload,
                        };
                        return Task.FromResult(response);
                    },
```

File: e2e/Tests/iothub/device/PayloadConventionE2eTests.cs [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+                    msg.TryGetPayload(out NjCustomPayload actual).Should().BeTrue();
+                    actual.StringProperty.Should().Be(payload.StringProperty);
+                    actual.GuidProperty.Should().Be(payload.GuidProperty);
+                    messageReceived.TrySetResult(true);
+                    var response = new DirectMethodResponse(200)
+                    {
+                        Payload = payload,
+                    };
+                    return Task.FromResult(response);
+                },
```
Comment (drwill-ms) [inline e2e/Tests/iothub/device/PayloadConventionE2eTests.cs:199]: tab in

File: e2e/Tests/iothub/device/PayloadConventionE2eTests.cs [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+            var messageReceived = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+            await deviceClient
+                .SetIncomingMessageCallbackAsync((msg) =>
+                {
+                    msg.TryGetPayload(out NjCustomPayload actual).Should().BeTrue();
+                    actual.StringProperty.Should().Be(payload.StringProperty);
+                    actual.GuidProperty.Should().Be(payload.GuidProperty);
+                    messageReceived.TrySetResult(true);
+                    return Task.FromResult(MessageAcknowledgement.Complete);
+                },
```
Comment (drwill-ms) [inline e2e/Tests/iothub/device/PayloadConventionE2eTests.cs:293]: tab in

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**